Update agent.mdx to improve memory explanation

rogerscuall opened the pull request:
Great job on this course, thanks for all your work.
The CodeAgent and its parent MultiStepAgent cannot be instantiated with a constructor while passing memory.  Memory is a field but not exposed in the constructor.  Also memory seems to be AgentMemory not a boolean.
In any case it would be a great to understand in this course more about the memory of smolagent.

sergiopaniego reviewed (commented):
Thanks for the improvement!

To enable memory in smolagents, we can set `reset=False` when using [run()](https://github.com/huggingface/smolagents/blob/05ccd4ce1f71db4b61c7a949247bf71c20a51dbd/src/smolagents/agents.py#L289). Would you like to update the second run call including that param?

`response2 = alfred_with_memory.run("...", reset=False)`

rogerscuall commented:
Hi @sergiopaniego ,
Thank you so much for looking into this and working on this course, it is very valuable.
I learned something new with your request, so let me know if I got it right:
```python
alfred_with_memory = CodeAgent(
    tools=[guest_info_tool, weather_info_tool, hub_stats_tool, search_tool], 
    model=model,
    add_base_tools=True,
    planning_interval=3
)
print(alfred_with_memory.run("My name is John"))
print(alfred_with_memory.run("What is my name?"))
```
Should return something like:
_**I'm sorry, but I don't have enough information to determine your name.**_
But if I do:
```python
alfred_with_memory = CodeAgent(
    tools=[guest_info_tool, weather_info_tool, hub_stats_tool, search_tool], 
    model=model,
    add_base_tools=True,
    planning_interval=3
)
print(alfred_with_memory.run("My name is John"))
print(alfred_with_memory.run("What is my name?", reset=False))
```
It should know that my name is Jonh, right?

Now I would like to provide you some feedback as a student of the course (aka external observer in this case).
I would like to see brief explanation of how the 3 agents implement the memory, in smolagent there is a decision to reset memory across runs, in llamaIndex an additional feature is needed "context", will the invoke method in langgraph caches the messages. I'm happy to contribute.

sergiopaniego reviewed (commented):
That's it! Thanks

rogerscuall commented:
I updated it I thought it would be great to ask the question of why 3 different frameworks decided to decouple the memory from agents, not attempting to answer it but just to invite for reflection.
I added a note each approach, I also added a link to LangGraph MemorySaver. 
I hope it helps.

sergiopaniego commented on units/en/unit3/agentic-rag/agent.mdx:
```diff
@@ -484,6 +483,11 @@ print(response['messages'][-1].content)
 </hfoptions>
 
+Notice that none of these three agent approaches directly couple memory with the agent. Is there a specific reason for this design choice 🧐?
+* smolagents: Memory is not preserved across different execution runs.
```
I'd rephrase this, saying that to preserve memory across runs, you must explicitly state it using `reset=False`

sergiopaniego commented on units/en/unit3/agentic-rag/agent.mdx:
```diff
@@ -484,6 +483,11 @@ print(response['messages'][-1].content)
+Notice that none of these three agent approaches directly couple memory with the agent. Is there a specific reason for this design choice 🧐?
+* smolagents: Memory is not preserved across different execution runs.
+* LlamaIndex: Requires explicitly adding a context construct for memory management within a run.
+* LangGraph: Offers options to retrieve previous messages or utilize a dedicated [MemorySaver](https://langchain-ai.github.io/langgraph/tutorials/introduction/#part-3-adding-memory-to-the-chatbot) component.
```
Since we're not explicitly using MemorySaver to keep things simple, I'd also add that idea in this line.

rogerscuall replied:
I'm not sure I fully understand your suggestion. Could you please propose a solution? Are you suggesting we explicitly mention MemorySaver as another option, or remove it altogether to keep things simple?

sergiopaniego replied:
It looks great as it is! :)

sergiopaniego commented on units/en/unit3/agentic-rag/agent.mdx:
```diff
@@ -484,6 +483,11 @@ print(response['messages'][-1].content)
 
+Notice that none of these three agent approaches directly couple memory with the agent. Is there a specific reason for this design choice 🧐?
+* smolagents: Memory is not preserved across different execution runs.
+* LlamaIndex: Requires explicitly adding a context construct for memory management within a run.
```
```suggestion
* LlamaIndex: Requires explicitly adding a context object for memory management within a run.
```

sergiopaniego reviewed (changes requested):
Thanks for the addition! Just a few ideas to improve it further! 😄

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/learn/agents-course/pr_409). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
